draft-ietf-jose-hpke-encrypt.md

```diff
@@ -191,41 +202,39 @@ the input MUST NOT be used and MUST be rejected.
+computed in Step 14 of {{encryption}}.
 
 ## Encapsulated Keys {#encapsulated-keys}
 
 HPKE encapsulated key is defined in {{Section 5 of I-D.ietf-hpke-hpke}}.
 
-In HPKE JWE Integrated Encryption, the JWE Encrypted Key of the sole recipient is the HPKE encapsulated key.
+In HPKE Integrated Encryption, the JWE Encrypted Key of the sole recipient is the HPKE encapsulated key.
 
-In HPKE JWE Key Encryption, each recipient JWE Encrypted Key is the encrypted content encryption key, and the value of JOSE Header parameter "ek"
-is base64url-encoded HPKE encapsulated key.
+In HPKE Key Encryption, each recipient's JWE Encrypted Key is the encrypted content encryption key, and the value of JOSE Header parameter "ek"
+is the base64url encoding of the HPKE encapsulated key.
 
-# Integrated Encryption
+# HPKE Integrated Encryption
 
-In HPKE JWE Integrated Encryption:
+When using HPKE Integrated Encryption:
 
-- The protected header MUST contain an "alg" that is JOSE-HPKE algorithm.
-- The protected header MUST contain an "enc" with value "int". This is an explicit exception to requirement in {{Section 4.1.2 of RFC7516}} that
-"enc" must be an AEAD algorithm. This is appropriate, as HPKE will perform plaintext encryption.
-- The protected header parameters "psk_id" MAY be present.
+- The protected header MUST contain an "alg" value that is JOSE-HPKE algorithm.
+- The protected header MUST contain an "enc" with the value "int".
+  This is an explicit exception to requirement in {{Section 4.1.2 of RFC7516}}
+  that "enc" must be an AEAD algorithm.
+  This is appropriate, as HPKE will perform plaintext encryption.
+- The protected header parameter "psk_id" MAY be present.
 - The protected header parameter "ek" MUST NOT be present.
 - There MUST be exactly one recipient.
-- The JWE Encrypted Key MUST be encapsulated key, as defined in {{Section 5 of I-D.ietf-hpke-hpke}}.
+- The JWE Encrypted Key MUST be the encapsulated key, as defined in {{Section 5 of I-D.ietf-hpke-hpke}}.
 - The JWE Initialization Vector and JWE Authentication Tag MUST be the empty octet sequence.
 - The JWE AAD MAY be present when using the JWE JSON Serialization.
 - The JWE Ciphertext is the ciphertext defined in {{Section 5.2 of I-D.ietf-hpke-hpke}}.
 - The HPKE info parameter contains the encoding of the Recipient_structure, which is described in
   {{recipient_structure}}.
-- The HPKE aad parameter MUST be set to the "Additional Authenticated Data encryption parameter", as specified in Step 14 of {{Section 5.1 of RFC7516}}.
-- Then follow Steps 11-19 of {{Section 5.1 of RFC7516}} (Message Encryption).
-
-When decrypting, the checks in {{Section 5.2 of RFC7516}},
-Steps 1 through 5 MUST be performed. The JWE Encrypted Key in Step 2 is the
-base64url-encoded encapsulated key.
+  The content_encryption_alg value in the Recipient_structure is "int".
+- The HPKE aad parameter MUST be set to the "Additional Authenticated Data encryption parameter", as specified in Step 14 of {{encryption}}.
```

Comment:
Per HPKE specification, using HPKE info and HPKE aad parameters simultaneously in single-shot SHOULD NOT be allowed. It really recommends returning an error, so using both will cause problems with implementations that follow this recommendation.

Reply:
@ilaril - In HPKE integrated encryption, it seems the only solution possible is to not set the the HPKE info parameter. The HPKE AAD will have to carry the "Additional Authenticated Data encryption parameter".